### DISCUSSION TITLE
Replacing cacheExchange with offlineExchange causes endless loading state

### DISCUSSION THREAD

**Discussion by RIP21** [urql-graphql/urql#1510] [Q&A]
I'm unsure if it's a bug, so let it be in the discussion first.
Here is my setup.

```typescript
import React, { useMemo } from 'react'
import { retryExchange } from '@urql/exchange-retry'
import { createClient, dedupExchange, errorExchange, fetchExchange, Provider } from 'urql'
import { devtoolsExchange } from '@urql/devtools'
import { refocusExchange } from '@urql/exchange-refocus'
import type { Cache } from '@urql/exchange-graphcache'
import { offlineExchange } from '@urql/exchange-graphcache'
import type { IntrospectionQuery } from 'graphql'
import { introspection } from 'packlets/generated'
import { AuthService } from 'services'
import { useAccessToken } from 'providers'
import { isNotEmptyOrNullish } from '@liveflow-io/utils-common'
import { makeDefaultStorage } from '@urql/exchange-graphcache/default-storage'

const storage = makeDefaultStorage({
  idbName: 'graphcache-v3', // The name of the IndexedDB database
  maxAge: 7, // The maximum age of the persisted data in days
})

export const UrqlProvider: React.FC = ({ children }) => {
  const token = useAccessToken()
  const client = useMemo(() => {
    const options: Parameters<typeof retryExchange>[0] = {
      initialDelayMs: 1000,
      maxDelayMs: 15000,
      randomDelay: true,
      maxNumberAttempts: 2,
      retryIf: (err) => !!err.networkError,
    }

    return createClient({
      requestPolicy: 'cache-and-network',
      url: '/graphql',
      exchanges: [
        dedupExchange,
        devtoolsExchange,
        refocusExchange(),
        offlineExchange({
          storage,
          schema: (introspection as unknown) as IntrospectionQuery,
          keys: {
             // ...
          },
          updates: {
            Mutation: {
              //...
            },
          },
        }),
        retryExchange(options),
        errorExchange({
          onError: (error) => {
            if (error.response?.status === 401) {
              AuthService.logout()
            }
          },
        }),
        fetchExchange,
      ],
      fetchOptions: () => {
        return {
          headers: { authorization: isNotEmptyOrNullish(token) ? `Bearer ${token}` : '' },
        }
      },
    })
  }, [token])

  return <Provider value={client}>{children}</Provider>
}

```
Ends up in an endless loading state with undefined data for the first queries route is too until unsubscribed and subscribed again, if the component then gets unmounted and remounted breaking the subscription to the state, it actually takes it correctly from the state.

In my case, whatever opened first (route) will have an endless loading state until you redirect to another route (that will work just fine) and redirect back to the page that was first opened.

This is how it looks.
![image](https://user-images.githubusercontent.com/3940079/113492456-bd50ca00-94d7-11eb-9c0a-b92ba08ace71.png)
****

No matter what I do, disable, enable different exchanges, etc., etc. it's broken and behaves very unpredictably.
This is the versions I'm using. 
```
    "@urql/core": "~2.0.0",
    "@urql/devtools": "~2.0.3",
    "@urql/exchange-auth": "~0.1.2",
    "@urql/exchange-graphcache": "~4.0.0",
    "@urql/exchange-refocus": "~0.2.2",
    "@urql/exchange-retry": "~0.2.0",
```
If I'll replace `offlineExchange` with `cacheExchange` (Graphcache) everything works just fine.

Fun fact that if I do that, one of the requests is that quick it never `console.log`s even a `fetching: true` goes straight to `data: {...}` but I guess it's only on localhost :) Is it possible? :D 

**UPD:**

Also, the query in this component for some reason never gets properly resolved if `offlineExchange` is used.

```typescript
import React, { useEffect } from 'react'
import { CurrentUserService, TrackingService } from '../utils'
import { useQuery } from 'urql'
import type { MeProvider_UserFragment } from './documents.generated'
import { MeProvider_MeDocument } from './documents.generated'

export const MeContext = React.createContext<{
  me?: MeProvider_UserFragment | null
  enabled: boolean
}>({
  enabled: false,
})

/**
 * This shall be used ONLY inside SecuredApp
 */
export const MeProvider: React.FC = (props) => {
  const [result] = useQuery({
    query: MeProvider_MeDocument,
  })

  useEffect(() => {
    if (result.data?.me) {
      CurrentUserService.setUserInfo(result.data.me)
      TrackingService.identify(result.data.me.id, result.data.me)
    }
    // eslint-disable-next-line
  }, [result.data?.me?.id])

  return (
    <MeContext.Provider
      {...props}
      value={{ me: result.data?.me, enabled: result.data?.enabled || false }}
    />
  )
}
```

If I do the same query from the outside and pass it down to this one as a prop it works as intended.
With `cacheExchange` it works just fine without any problems.

Tree is like following.

`Root > UrqlProvider (at the top of the post) > Router > SecuredApp > MeProvider (right above) > {CurrentyOpenedRoute}`

**Comment by RIP21** [comment]
@kitten aaand another weird behavior.

The setup above (with cacheExchange), a scenario like that.
```
  const [kekResult, refetchKek] = useQuery({
    query: `query { kek }`,
  })

  useEffect(() => {
    setInterval(() => {
      refetchKek({
        requestPolicy: 'cache-and-network',
      })
    }, 5000)
  }, [])
```

Causes two `graphql` calls each 5 seconds (first render single request) one is getting canceled and second resolves.

Also, it triggers the `loading` state even tho `cache-and-network` should set `stale:true` only in this case, it actually goes to the `fetching` state (cannot detect that in  `console.log` only `debugger` stops inside `if(result.fetching)`, to avoid spinner I decided to give a try to Suspense (quite cool!)

Any ideas what can cause that?
And I'm sure that the interval above, as well as the effect, runs only once, I checked that with the debugger.

I'm pretty sure I also have only one variant of `urql` and other exchanges in my monorepo `/@urql/core/2.0.0_graphql@15.5.0:`
and `/urql/2.0.1_graphql@15.5.0+react@17.0.1:` and one instance of `react 17.0.1` so I guess it's not caused by some weird dependencies conflict or circular dependency or whatever.
it just getting called twice on `refetch`.

**Reply by JoviDeCroock** [reply]
Are you running `StrictMode`? Also have you tried with `network-only`, `cache-and-network` sounds a bit counterintuitive for a refetch.

**Reply by RIP21** [reply]
I'm running in the StrictMode yes.
The idea is to silently update the cache.

**Reply by kitten** [reply]
That's normal. The first will always be probing for synchronous state, the second will be resulting from the next render cycle and will actually be allowed to run through. You'll only see `stale` and a result if there's a result to show in the first place. If there isn't then it won't and will fetch.

We've also noticed that you've set `cache-and-network` as the default policy. This may seem like a good idea at first but we actually don't recommend this, especially in combination with a normalised cache. The reason for that is that it indicates that you're afraid of stale state, and the point of a normalised cache actually already counteracts that. A normalised cache's job is to remain up-to-date in the presence of changes being only caused by mutations.

That's why we've got the `requestPolicyExchange`, because we don't want to always be over fetching and refetching on each call.

The other part is that you may also not want to run this policy on an interval and maybe you'd instead want to use `network-only`, because it will set `fetching` but your hook will still remember its last state.

What we're mostly concerned about is that you have an account query that can be launched from multiple parts of your application. This may be using `cache-and-network` which can cause subsequent queries to wait for the network result rather than actually using the cached result. In other words, it's not concurrent-safe.

**Reply by RIP21** [reply]
@kitten sorry, the config one above is for offline.
When I'm using Graphcache alone (as offline is broken for me)
I have default requestPolicy of `cache-first`
And only in this scenario of polling I use `cache-and-network`.

**Reply by RIP21** [reply]
So to reiterate what's my issue in this thread (unrelated to offline mode issue above)
- I have default `cache-first`.
- I open a page that only has this query + refetch effect.
- On the initial render it fires query nicely and fetches data.
- Then I want it to update every 5 seconds.
- Interval calls `refetchKek` and it leads to 2 requests instead of one, one gets cancelled immediately
- A quick re-render of Spinner => Content (barely visible by eye), which is unexpected as `fetching` should not be `true` as it's a `cache-and-network` refetch so `stale` flag should be `true` while fetching, but not `fetching`. E.g. I want to achieve silent cache/state change without spinners.

**Reply by RIP21** [reply]
Funnily enough in another component, I have a similar setup with two different queries that are refetching every 3 seconds and they do not cause anything like that. And if I paste this one over close to them it's all good.
If I move it to some lightweight stupid page like NotFound, it starts to do these double calls.
And another funny thing that if you have code like:
```ts
  if (kekResult.fetching) {
    console.log("Fetching")
  }
```
It's not outputs to the console.
But 
```ts
  if (kekResult.fetching) {
    debugger
  }
```
Actually stops on refetch and you can clearly see that `kekResult` has `data: {...}` and `fetching:true` I don't know what kind of voodoo magic is that, but this is how it is.

**Reply by RIP21** [reply]
@kitten @JoviDeCroock will be happy to jump on the call and show that all on the real code sharing screen. As this is SUPER weird :)

**Reply by RIP21** [reply]
Ok, so the solution it seems is to use `network-only` when doing refetches and it will be a hidden refetch, and it will refresh data in cache without setting `fetching: true`. So this one is sorted. Offline mode is not tho.

**Comment by RIP21** [comment]
@kitten @JoviDeCroock Any ideas related to `offlineCache` exchange issues tho? :) As I really want to get my hands dirty with offline mode, but simply can't due to these issues.

I can try to create a reproduction if no ideas you folks are having if this will help (it may take a lot of time), but as I said above, I'll be glad to show everything on a call on the real codebase.

**Reply by kitten** [reply]
We can probably narrow this down if you could post us the log output of the `debugExchange` placed in front of the `offlineExchange`. Apart from that a reproduction would be quite helpful ✌️

**Reply by RIP21** [reply]
@kitten will do once will ship some very important features :) This time with `suspense: true` so may change behavior a little.
But I still haven't refactored everything to suspense, so will be able to give you logs for both (maybe suspense mode will "magically" fix all the issues)

**Reply by RIP21** [reply]
@kitten Okay, it's hard to output it all as it's just not possible to copy-paste nor export somehow.
So here is the screenshot.
This is offlineExchange + suspense: false, networkPolicy: cache-and-network
Operation with key 198 starts multiple times but never gets completed for some reason so endless spinner is shown and it's stuck on endless loading.

<details>
  <summary>Screenshot,  offlineExchange + **suspense: false**, networkPolicy: cache-and-network (in client)</summary>
  
![image](https://user-images.githubusercontent.com/3940079/114037228-11b1cc00-9881-11eb-83b3-409f115b5b11.png)
  
</details>


Same but **suspense: true**
<details>
  <summary>Screenshots of,  offlineExchange + **suspense: true**, networkPolicy: cache-and-network. (in client)</summary>
  
![image](https://user-images.githubusercontent.com/3940079/114037979-c1873980-9881-11eb-8d47-a3cd850965ec.png)
Continuation:
![image](https://user-images.githubusercontent.com/3940079/114038116-e085cb80-9881-11eb-8215-478ce266434f.png)
  
</details>


The only issue in suspense mode with `offlineExchange` is that for some reason it gets to `fetching` state somehow?! And my safety net of
```typescript
  if (result.fetching) {
    // NOTE: Suspense should never let us get here! Woohoo!
    throw new Error('Is suspense mode deactivated?')
  }
```
Actually, triggers which is weird as in suspense mode we shouldn't get into that ever. Although I see that my fallback loading states in many places are actually working fine.

**Reply by kitten** [reply]
I think as said with `cache-and-network` and `executeQuery` it's quick to get into invalid states. If that's actually the issue, ie related to duplicate queries and the request policy, changing that would fix it, but those seem to be hard for you to pin down?

If that's the cause this PR may help: https://github.com/FormidableLabs/urql/pull/1515
But it's experimental and I haven't tested it yet. But you could give it a try by using `@urql/core` as built by CodeSandbox CI

**Reply by RIP21** [reply]
@kitten above is not related to retries/executeQuery and `cache-and-network`. It's related to the main question e.g. `offineExchange` that is broken probably.
I'm doing as per documentation.
E.g change `cacheExchange` with `offlineExchange` + add storage + set default networkPolicy to `cache-and-network`.
And logs above are related to that.

**Reply by RIP21** [reply]
In suspense mode it's somewhat working but gets into `fetching` state when it shouldn't.
While without suspense mode it's completely broken.
Just `useQuery` no explicit usage of `executeQuery` etc. Just components that are getting rendered and are calling `useQuery` inside them.
Regular cache all works perfectly, `offlineExchange` - all broken (without suspense especially, with it just fetching flag is getting to `true` somehow which's concerning)

**Reply by kitten** [reply]
@RIP21 what I was referring to was that a default request policy of `cache-and-network` isn't recommended because it could be causing these issues especially if the app runs duplicate queries and if it uses a lot of paused/executed queries. This is especially my suspicion because it goes away in Suspense mode, which can prevent some of these problems because of its secondary suspense cache

**Reply by RIP21** [reply]
@kitten okay then why in the docs there is a recommendation to add `cache-and-network`?
https://formidable.com/open-source/urql/docs/graphcache/offline/#offline-behavior
Plus endless loading is appearing when suspense more is `false`, that's a completely broken setup (this combo  offlineExchange + **suspense: false**, networkPolicy: cache-and-network (in client) as default)
With `suspense` it's somewhat working but gets into `fetching:true` state, and this is maybe because it's `cache-and-network` that is on.

And If `cache-and-network` is broken/unstable, why it's in the `urql` in the first place? :)

**Reply by kitten** [reply]
@RIP21 It's not broken and unstable, but comes with some footguns, because it's not meant to be the default. We'll update that part of the docs but mainly Offline still has some things that will need improvements. One of the important parts here is that it's meant to say that `cache-and-network` needs to be used _selectively_ to ensure that persisted data is updated at all. If you think about what's going on in an Offline-ready app, at some point it will have collected enough data to _always_ read from cache, so it'll need hints to know when to refetch.

That's why we have the `requestPolicy` exchange. It's a first attempt at showing how _selectively_ some operations can be upgraded to refresh the cache. What we didn't anticipate is that a lot of apps and teams started _overusing_ cache-and-network.

If we relate this to duplicate queries; similarly it should be extremely rare to duplicate queries. And if you do you can see how `cache-and-network` runs into trouble. If one query starts and receives a cached result, then if a duplicate query is mounted it can't get a cached result and instead loads (like it encountered a cache miss) which is because `dedupExchange` prevents two operations from running simultaneously.

That on its own isn't an issue, but should explain to you why Suspense mode may change this behaviour. The endless loading loop is likely caused by some input change, remount/unmount issue, or effect triggering an imperative reload when this duplication happens in your app. That's non recoverable. However Suspense mode is probably preventing this state from being visible in your app, because the duplicate query triggers a suspense loading switch, where that logic isn't being triggered. So you see, there's two things that combine in an unfortunate manner in your app.

I can't tell you about the exact cause obviously, since I don't have code or a case to look into, but this is what this sounds like, since we've seen this before. So that's why `cache-and-network` comes with some gotchas.

That's why I've drafted: https://github.com/FormidableLabs/urql/pull/1515. Because frankly, I don't want to deal with these mistakes anymore and am searching for a remedy for duplicate queries 😅 So to summarise, I don't know whether #1515 works yet, because I haven't tested it yet due to a lack of an environment / sandbox that minimally presents this behaviour. We had one in the past, but none that exactly exhibits this edge case. However, you can give it a try by using the CodeSandbox CI build: https://ci.codesandbox.io/status/FormidableLabs/urql/pr/1515/builds/116282

You'll have to add a Yarn resolution for `@urql/core` to this CSB CI build's package URL, e.g. `"@urql/core": "https://pkg.csb.dev/FormidableLabs/urql/commit/45388120/@urql/core"`

**Reply by RIP21** [reply]
Ok, that's make sense then.
Will check this PR you have, will report back on results.

**Reply by RIP21** [reply]
@kitten 
With this PR applied I can confirm that all issues for "`offlineExchange` + `suspense: true`, `networkPolicy`: cache-and-network (in client) as default" was solved. In particular `fetching: true` is never getting set to `true` so my safe net `if` in the root of the app
```ts
  if (result.fetching) {
    // NOTE: Suspense should never let us get here! Woohoo!
    throw new Error('Is suspense mode deactivated?')
  }
```
is never called.

Regarding the same setup but without `suspense` 
It's still broken. Endless loading state for certain first queries. I'll try to build a repro on CodeSandbox so you can investigate.
Important to note that  "`offlineExchange` + `suspense: false` BUT `networkPolicy`: **cache-first** (in client) as default" haven't fixed endless loading. So it seems it's not connected to `cache-and-network` footguns/problems/whatever that you mentioned above but to something else.

I will attach the repro as soon as I'll get one.

**Reply by RIP21** [reply]
@kitten additional small question.
When I invalidate certain pieces in Graphcache + Offline Mode will it be invalidated from DB too?
Or Urql will fetch new data and will update DB instead? How does it work?

**Reply by kitten** [reply]
Yes, when you call `cache.invalidate` then the storage will see that as an update to its entries; specifically deleted entries will be marked as `undefined` rather than a serialised value. The `default-storage` although its TTL-based covers this

**Reply by RIP21** [reply]
@kitten hey, just sent you an invite to private repo with reproduction.
https://github.com/RIP21/urql-offline-issues-repro

It's using our dev environment :) And it's just my original monorepo from which I quickly stripped all useless info.
You should be able to login using your e-mail or Google (it's Auth0 auth) and then get "onboarded".
And then once you will be redirected to /integrations page you need to refresh the page to see endless loading.

App code is in `ui/app`

Queries that will be listened to on the integration page are in
`pages/secured/SecuredApp`
`pages/secured/Integrations`

UrqlProvider and client creation is in `packlets/providers/UrqlProvider`

Any questions, let me know.

And yeah, how to run it, check `README.md` 

If you would like to try your commit version of the `@urql/core` you can go to 
`common/config/rush/pnpmfile.js`

And add this into `readPackage` function
```
  changeDepGlobal(
    '@urql/core',
    'https://pkg.csb.dev/FormidableLabs/urql/commit/45388120/@urql/core',
    true,
  )
```
And run `rush update --full`

**Comment by baransu** [comment]
Actually I'm facing a similar problem. Using `cacheExchange` + `suspense:false` + `cache-first` (default request policy). I'm making the first query on the initial render (it hits cache and everything is returned correctly) which renders 3 components that are making 3 separate queries (all of them with `cache-first`) but non of them returns data and is stuck in infinite `fetching:true` state. When the cache is cold everything works as expected and queries hit the server because of a cache miss.

I'm attaching devtools events screenshot that shows that the last 3 queries never receive a response (neither from cache nor network). The second query is introspection query triggered by devtools exchange.

![Screenshot 2021-05-29 at 01 25 51](https://user-images.githubusercontent.com/9558691/120050202-daf46880-c01c-11eb-8649-022c3ec07c75.png)

After making an experiment with delaying 1st query by fixed 1000ms by setting `pause:true` and triggering re execute (with default `cache-first` request policy) inside useEffect with 1000ms timeout everything looks to work as expected as well.

**Reply by bryansum** [reply]
Just ran into a similar issue, traced it back to a combination of `offlineExchange` + usage of `storage`

**Reply by sarink** [reply]
I'm having this exact problem, @baransu @bryansum @kitten is there a solution? These queries ought to be getting deduped and/or read from the cache, but instead they wind up fetching indefinitely.
